src/CloudfrontStack.ts

```diff
@@ -0,0 +1,253 @@
+      defaultBehavior: {
+        origin: s3Origin,
+        viewerProtocolPolicy: ViewerProtocolPolicy.HTTPS_ONLY,
+        functionAssociations: [
+          {
+            function: redirectFunction,
+            eventType: FunctionEventType.VIEWER_REQUEST,
+          },
+        ],
+      },
+      errorResponses: this.errorResponses(),
+      logBucket: this.logBucket(),
+      minimumProtocolVersion: SecurityPolicyProtocol.TLS_V1_2_2021,
+    });
+
+
+    this.addDnsRecords(distribution, projectHostedZoneId, projectHostedZoneName);
+
+    this.addPublicBuckets(props.configuration, distribution);
+
+    //export for importing the distribution in the storageStack
+    new StringParameter(this, 'cf-domain', {
+      stringValue: distribution.distributionDomainName,
+      parameterName: Statics.ssmCloudfrontdomainName,
+    });
+    new StringParameter(this, 'cf-id', {
+      stringValue: distribution.distributionId,
+      parameterName: Statics.ssmCloudfrontDistributionId,
+    });
+
+  }
+
+  private errorResponses() {
+    const errorCodes = [403, 404, 500];
+    return errorCodes.map(code => {
+      return {
+        httpStatus: code,
+        responseHttpStatus: code,
+        responsePagePath: `/http-errors/${code}.html`,
+      };
+    });
+  }
+
+
+  private addPublicBuckets(configuration: Configuration, distribution: Distribution) {
+    const customCachePolicy = new CachePolicy(this, 'ThreeMonthCachePolicy', {
+      cachePolicyName: 'ThreeMonthCachePolicy',
+      defaultTtl: Duration.seconds(60 * 60 * 24 * 90), // 3 months
+      minTtl: Duration.days(1),
+      maxTtl: Duration.days(365),
+      enableAcceptEncodingGzip: true,
+      enableAcceptEncodingBrotli: true,
+    });
+
+
+    // Retrieve KMS key and add CloudFront access policy
+    const key = Key.fromKeyArn(this, 'EncryptionKey',
+      ssm.StringParameter.valueForStringParameter(this, Statics.ssmGeoStorageKmsKeyArn));
+
+    // Create custom resource to add policy to KMS key allowing CloudFront to decrypt
+    new CloudfrontKmsPolicy(this, 'KmsPolicy', {
+      kmsKey: key,
+      cloudfrontDistributionArn: distribution.distributionArn,
+    });
+
+    for (const bucketSettings of configuration.buckets) {
+      if (bucketSettings.cloudfrontBucketConfig && bucketSettings.cloudfrontBucketConfig.exposeTroughCloudfront) {
+        // Use a unique ID for each bucket to avoid conflicts
+        const uniqueId = `${bucketSettings.cdkId}-cf`;
+        const bucket = Bucket.fromBucketAttributes(this, uniqueId, {
+          bucketName: bucketSettings.name,
+          encryptionKey: Key.fromKeyArn(this, `${uniqueId}-key`, ssm.StringParameter.valueForStringParameter(this, Statics.ssmGeoStorageKmsKeyArn)),
+        });
+
+        const s3Origin = S3BucketOrigin.withOriginAccessControl(bucket, {
+          originAccessLevels: [AccessLevel.READ, AccessLevel.LIST],
+        });
+
+        this.addBucketPolicyForCloudfront(bucket, distribution);
```

Comment:
Omdat het anders niet wilde lijkt dit me nu prima, maar het hele doel van de origin access control is (bucket)toegang via cloudfront, dus ik wil hier nog wel een keer goed naar kijken. Ik zou dus wel een issue willen aanmaken om die ontsluiting via cloudfront nog 's goed te bekijken.